Create migration guide for SDKv5 to v6 (#1661)
LiskHQ/lisk-docs: modules/ROOT/pages/references/migration.adoc

= SDK Migration Guide
Mona Bärenfänger <mona@lightcurve.io>
// Settings
:toc: preamble
:idprefix:
:idseparator: -
:docs-general: ROOT::
:docs-lisk-v5: lisk-sdk::
:docs-understand-modules: {docs-general}understand-blockchain/sdk/modules-commands.adoc
:sectnums:
// URLs
:url_ethereum_events: https://medium.com/mycrypto/understanding-event-logs-on-the-ethereum-blockchain-f4ae7ba50378
:url_wiki_pubsub: https://en.wikipedia.org/wiki/Publish%E2%80%93subscribe_pattern
:url_lip_37: https://github.com/LiskHQ/lips/blob/main/proposals/lip-0037.md
// Project URLs
:url_sdkv5_rpc_events: {docs-lisk-v5}events.adoc
:url_api_rpc_getEvents: {docs-general}api/lisk-node-rpc.adoc
:url_standardevent: {docs-understand-modules}#standard-event
:url_methods: {docs-understand-modules}#methods
:url_endpoints: {docs-understand-modules}#endpoints

> :url_metadata: {docs-understand-modules}#metadata

Reviewer: ```suggestion
:url_metadata: {docs-understand-modules}#module-metadata
```

.Main benefits of this change
[%collapsible]
====
Increased rewards for validators::
By doing this, validators will still be rewarded for executing the logic of the transaction until the point where it failed.
Increased security::
Additionally, it mitigates the danger of DDoSing sidechains by spamming transactions that will fail, because the transaction fee has to be paid in any case.
====

To verify, if a transaction that is included in a block was executed successfully, Lisk SDK v6 introduces the <<addition-of-blockchain-events>>.

== Addition of blockchain events

**Blockchain events** are **newly introduced**, and **Module pub/sub events** are **completely removed** in Lisk SDK v6.

Although similar in name, Blockchain events follow a completely different approach.
To avoid confusion between the two events, please read the summary below, which is comparing the new and old events inside modules.

.Main benefits of this change
[%collapsible]
=====
Required, if failed transactions are included in blocks::
As explained above, <<failed-transactions-are-included-in-blocks>> in v6.
+
[CAUTION]
====

> This means, if you wish to verify that a transaction was successfully executed, it is not sufficient anymore to check, if it is included in a finalized block.

Reviewer 2: the statement sounds a bit unclear, maybe it can be simplified

Author: I tried to simplify it now, and removed the CAUTION admonition as well

Additional events can be defined per module by the sidechain developer.
=====

[tabs]
=====
Blockchain events - SDKv6::
+
--
[cols="1h,6"]
|===
|Description
|Blockchain events are introduced newly in Lisk SDK v6.
Blockchain events are still defined and emitted inside the module, but it is not possible to subscribe to them via the RPC API of a node.

Instead, blockchain events are **logged per block**, i.e. directly included in the block header and can be queried through the RPC API of a node by requesting the RPC endpoint `chain_getEvents`.

*Blockchain events in Lisk SDK v6 are following a concept comparable to the {url_ethereum_events}[Ethereum event log^].*

Besides the newly introduced blockchain events, there is still a couple of xref:{url_understand_rpc_events}[RPC events] included in the Lisk Framework, which can be retrieved via public-subscribe, as in v5.

The sidechain developer can define new blockchain events per module as desired, but the RPC events are not customizable any more.
Only new blockchain events can be included in the sidechain client by the developer.

|Purpose

> |Blockchain events are a way for modules to store important information which is not included in the transactions or block assets, directly inside the block header.

Reviewer 2: ```suggestion
|Blockchain events are a way for modules to store important information which is not included in the transactions or block assets, verifiable using eventRoot property present inside the block header.
```

They are part of the overall state of the blockchain, as an event root of all events included in a particular block is stored in the block header.

They can include a lot of additional data, if required, as the events themselves can be removed from the stores of the node after a certain time, and therefore don't "pollute" the blockchain itself.

|Definition
a|
[source,js]
----
public constructor() {
    super();
    // registration of stores and events
    this.events.register(NewHelloEvent, new NewHelloEvent(this.name));
}
----

> For more information how to create the corresponding event class, please check out the guide xref:{url_build_module_event}[]

Author: ```suggestion
For more information on how to create the corresponding event class, please check out the guide xref:{url_build_module_event}[]
```

An event informs services that subscribed to it, if a certain event happened (e.g. a transaction was executed), and often contain additional data, providing more information or context about the event.

*Module events in Lisk SDK v5 follow the {url_wiki_pubsub}[publish-subscribe-pattern^].*

|Purpose
|Events are used to communicate about certain events in real time, and to prevent reoccuring RPC request, just to check if there are any changes.

|Definition
a|
[source,js]
----
public events = ['newHello'];
----

For more information how to create the corresponding event class, please check out the guide xref:{url_build_module_event}[]

|Publishing
a|
[source,js]
----
this._channel.publish('hello:newHello', {
    sender: tx.senderAddress.toString('hex'),
    hello: helloAsset.helloString
});
----

|Retrieving
a|
In Lisk SDK v5, events could be subscribed directly via the API client.

If an event was missed, there was no way of retrieving the event, after it was emitted by a node.

[source,js]
----
client.subscribe('app:block:new', ( data ) => {
  console.log('new block:',data);
});
----
|===
--
=====

== Chain ID replaces network ID

> Chain identifiers (or chain IDs), for transaction signatures and block signatures, were already introduced in Lisk SDK v5 as network identifiers to prevent replay attacks on other chains.

Reviewer 2: we can also mention to identify chains for making cross chain transfers in the ecosystem

[tabs]
=====
Chain ID example::
+
--
 00000000

--
Network ID example::
+
--
 4c09e6a781fc4c7bdb936ee815de8f94190f8a7519becd9de2081832be309a99

--
=====

Chain identifiers are 4-byte values that follow a specific format:
the first byte is used to identify the network in which the chain is running (either the Lisk Mainnet, Lisk Testnet, or any other test network);
the other 3 bytes identify the sidechain within the network.

NOTE: The network-specific prefix is included explicitly to ensure that a chain does not use the same chain identifier in the test network as in the mainnet.

.Main benefits of this change
[%collapsible]
====
Improved developer experience::
* The chain identifier can be directly set by the blockchain creator, which is more convenient than generating a random 32-byte value.
Improved user experience::
* By using a much shorter ID, users can easily verify that they are signing a transaction for the correct blockchain.
====

== Module & command IDs are removed

The module and command IDs are removed completely in Lisk SDKv6.

Module and command name are used as unique identifiers for modules and commands instead.

.Main benefits of this change
[%collapsible]
====

> Improved developer experience::

Author: ```suggestion
Enhanced developer experience::
```

    "id": "1",
    "method": "hello_getHello",
    "params": {
        "address": "lskuz5p98kz3mqzxnu68qdrjxtvdvr2o7pprtj4yv"
    }
}'
----

[source,typescript]
----
// How to invoke an endpoint that needs some data input.
const data = await client.invoke('namespace_endpointName', input);
console.log(data);
----
|===
--
Actions - SDKv5::
+
--
[cols="1h,6"]
|===

|Description
|An interface between a module and an external system via an RPC endpoint.

|Definition
a|
[source,typescript]
----
export class HelloModule extends BaseModule {

    // ...

    public actions = {
        amountOfHellos: async () => {
            const res = await this._dataAccess.getChainState(CHAIN_STATE_HELLO_COUNTER);
            const count = codec.decode(
                helloCounterSchema,
                res
            );
            return count;
        },
    };

    // ...
}
----

|Usage
a|
[source,typescript]
----
const data = await client.invoke('app:actionName', input);
----
|===
--
=====

== Migration of store from v5 to v6

A store, aka key-value store is a special kind of database which follows a data storage paradigm designed for storing, retrieving, and managing associative arrays.

> In the Lisk SDK, stores are used to store the on-chain and off-chain related data of a node persistent.

Reviewer: ```suggestion
In the Lisk SDK, stores are used to store the on-chain and off-chain related data of a node.
```

        // 1. Get message store
        const messageSubStore = this.stores.get(MessageStore);
        // 2. Get the address from the endpoint params
        const { address } = ctx.params;
        // 3. Validate address
        if (typeof address !== 'string') {
            throw new Error('Parameter address must be a string.');
        }
        cryptography.address.validateLisk32Address(address);
        // 4. Get the Hello message for the address from the message store
        const helloMessage = await messageSubStore.get(
            ctx,
            cryptography.address.getAddressFromLisk32Address(address),
        );
        // 5. Return the Hello message
        return helloMessage;
    }
}
----
|===

== Addition of metadata

Lisk SDK v6 introduces a new RPC endpoint to get all existing metadata related to a node.
This includes metadata of all the modules which are registered on the node.

The module developer can now easily define which data should be returned by the endpoint for the particular module by adjusting the newly introduced xref:{url_metadata}[metadata] method.

.Main benefits of this change
[%collapsible]
====
Improved user experience::

> * Users can now query all relevant metadata about a module in a consistent manner.

Reviewer: ```suggestion
* Users can now query all the relevant metadata about a module in a consistent manner.
```

}
----

> For more information on how to create the corresponding event class, please check out the guide xref:{url_build_module_event}[]

Reviewer: ```suggestion
For more information on how to create the corresponding event class, please check out the guide on xref:{url_build_module_event}[]
```

Author: In my ears it doesn't sound right here to add an `on`, because the link label coming afterwards resolves to the page title of the guide: "How to create a blockchain event", so imho the sentence should be

- ..., please check out the guide "How to create a blockchain event".

Author: btw I also updated the terminology changes to include all pos related terminologies, please check

Reviewer: The following three are correct, so you can use either one of them.

please check out the guide "How to create a blockchain event".
please check out the guide: How to create a blockchain event.
please check out the guide on How to create a blockchain event.

Author: so in that case, we agree my original sentence is correct?

Reviewer: Nope, coz the original sentence will be like this:

- please check out the guide How to create a blockchain event.

Author: Please be aware that the guide title is rendered as a link.

Author: ```suggestion
For more information on how to create the corresponding event class, please check out the guide: xref:{url_build_module_event}[]
```
